### DISCUSSION TITLE
Detecting RabbitMQ version after uninstall.

### DISCUSSION THREAD

**Discussion by inikulshin** [rabbitmq/rabbitmq-server#12064] [Questions] [answered]
Is it possible to detect RabbitMQ version from the files in RABBITMQ_BASE (mnesia? schema.dat?) after RabbitMQ server is uninstalled?

Erlang is installed, if it helps.

I want to add this check before installation to prevent downgrade (in rollback, after failed upgrade) and potential data corruption.

**Comment by michaelklishin** [comment, marked as answer]
No, not really. The on-disk formats do not change very often and the data directory in general does not have any versioning (individual files, namely Raft log segments, might be).

The best you can do is to store version information before the node is stopped and uninstalled:

``` shell
rabbitmq-diagnostics status --formatter=json | jq .rabbitmq_version
```

**Reply by inikulshin** [reply]
Hm, but how RabbitMQ server knows that it need to make any db upgrades on first startup after RabbitMQ upgrade?

**Reply by michaelklishin** [reply]
Previously there was a migration function mechanism that was more or less a sequential list of function to run.

These days it's mostly feature flags and different parts of the code detecting different loaded data formats. Nothing depends on a particular version number. A feature flag, when enabled, can and usually does run a migration function.
